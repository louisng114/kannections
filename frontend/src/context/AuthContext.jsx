import React, { createContext, useState, useEffect } from 'react';
import { jwtDecode } from "jwt-decode";
import KannectionsApi from '../helpers/api';

const AuthContext = createContext();

// functions for user authentication
const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  useEffect(() => {
    const storedToken = localStorage.getItem('token');
    if (storedToken) {
      KannectionsApi.token = storedToken;
      fetchUserFromToken(storedToken);
    }
  }, []);

  const fetchUserFromToken = async (token) => {
    try {
      const decoded = jwtDecode(token);
      const user = await KannectionsApi.getUser(decoded.username);
      setUser(user);
    } catch (error) {
      console.error('Failed to decode token:', error);
    }
  };

  const signup = async (username, password) => {
    try {
        const token = await KannectionsApi.register({ username, password });
        localStorage.setItem('token', token);
        KannectionsApi.token = token;
        fetchUserFromToken(token);
    } catch (error) {
      console.error("Registration failed:", error);
      throw error;
    }
  };

  const login = async (username, password) => {
    try {
      const token = await KannectionsApi.getToken({ username, password });
      localStorage.setItem('token', token);
      KannectionsApi.token = token;
      fetchUserFromToken(token);
    } catch (error) {
      console.error("Login failed:", error);
      throw error;
    }
  };

  const logout = () => {
    localStorage.removeItem('token');
    setUser(null);
    KannectionsApi.token = null;
  };

  return (
    <AuthContext.Provider value={{ user, signup, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
};

export { AuthProvider, AuthContext };
