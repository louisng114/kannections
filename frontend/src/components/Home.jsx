import { useContext } from 'react';
import { AuthContext } from '../context/AuthContext';

const Home = () => {
    const { user } = useContext(AuthContext);

    const description = (
        <p>
          <strong>KANnections</strong> is a <a href="https://www.nytimes.com/games/connections">Connections</a>-inspired game for reviewing Kanji! Click "Game" on the navigation bar and select a JLPT level to start a game!
        </p>
    );

    if (!user) {
    return (
        <div>
        <h1>
            WELCOME TO <ruby>æ¼¢<rt>KAN</rt></ruby>NECTIONS!
        </h1>
        {description}
        </div>
    );
    }
    return (
    <div>
        <h1>WELCOME BACK, {user.username}!</h1>
        {description}
    </div>
    );
};

export default Home;