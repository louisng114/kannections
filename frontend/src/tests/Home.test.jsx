import { render } from '@testing-library/react';
import Home from '../components/Home';
import { AuthProvider } from '../context/AuthContext';

test('renders without crashing', () => {
  render(
    <AuthProvider>
        <Home />
    </AuthProvider>
  );
});

it("matches snapshot", function() {
  const {asFragment} = render(
      <AuthProvider>
        <Home />
      </AuthProvider>
    );
  expect(asFragment()).toMatchSnapshot();
});
