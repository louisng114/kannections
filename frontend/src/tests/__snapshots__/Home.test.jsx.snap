// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h1>
    WELCOME TO 
    <ruby>
      æ¼¢
      <rt>
        KAN
      </rt>
    </ruby>
    NECTIONS!
  </h1>
</DocumentFragment>
`;
